README.md
<h1 align="center">morfy</h1>
<h3 align="center">Fast morphing library for the web</h3>
<p align="center">
  <a href="https://www.npmjs.com/package/morfy" target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/morfy.svg">
  </a>
  <a href="https://github.com/TimoBechtel/morfy/blob/main/LICENSE" target="_blank">
    <img alt="License: MIT" src="https://img.shields.io/github/license/TimoBechtel/morfy" />
  </a>
</p>
<p align="center">
  ·
  <a href="https://github.com/TimoBechtel/morfy#readme">Homepage</a>
  ·
  <a href="https://timobechtel.github.io/morfy/">View Demo</a>
  ·
  <a href="https://github.com/TimoBechtel/morfy/issues">Report Bug / Request Feature</a>
  ·
</p>

## Table of Contents

- [About](#About)
- [Installation](#Install)
- [Usage](#usage)
- [Test](#run-tests)
- [Contact](#contact)
- [Contributing](#Contributing)
- [License](#license)

## About

Morfy allows you to morph one HTML Element into another. It uses CSS transforms, so it's really fast and animations are cancelable. It is also toggleable (funny word, huh?). You can use it for example to morph a button into a modal.

## Install

### NPM:

```sh
npm install morfy
```

## Usage

### As module:

```javascript
import { createMorphable } from 'morfy';
```

Then:

```javascript
const button = document.getElementById('test-button');
const modal = document.getElementById('modal');
const closeButton = document.getElementById('modal-close-button');

const morphable = createMorphable(button, modal, {
  timingFunction: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
});

button.addEventListener('click', () => {
  morphable.morph();
});

closeButton.addEventListener('click', () => {
  morphable.revert();
});
```

Docs:

```typescript
interface MorfyOptions {
  /**
   * duration in seconds
   */
  duration: number;
  /**
   * css timing function
   * provides autocompletion when using typescript
   * @example 'ease-in'
   */
  timingFunction: string;
  /**
   * css properties to be transitioned from source to target
   * provides autocompletion when using typescript
   */
  affectedCssProperties: string[];
}

interface Morphable {
  /**
   * start morphing
   */
  morph: () => void;
  /**
   * morph to initial state
   */
  revert: () => void;
}

/**
 * Creates and initializes morphable object
 */
function createMorphable(
  source: HTMLElement,
  target: HTMLElement,
  options: MorfyOptions
): Morphable;

/**
 * Directly morph things
 */
function morph(
  source: HTMLElement,
  target: HTMLElement,
  options: MorfyOptions
): void;
```

## Run tests

```sh
npm run test
```

## Contact

👤 **Timo Bechtel**

- Website: https://timobechtel.com
- Twitter: [@TimoBechtel](https://twitter.com/TimoBechtel)
- GitHub: [@TimoBechtel](https://github.com/TimoBechtel)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />

1. Check [issues](https://github.com/TimoBechtel/morfy/issues)
1. Fork the Project
1. Create your Feature Branch (`git checkout -b feat/AmazingFeature`)
1. Test your changes `npm run test`
1. Commit your Changes (`git commit -m 'feat: add amazingFeature'`)
1. Push to the Branch (`git push origin feat/AmazingFeature`)
1. Open a Pull Request

### Commit messages

This project uses semantic-release for automated release versions. So commits in this project follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/) guidelines. I recommend using [commitizen](https://github.com/commitizen/cz-cli) for automated commit messages.

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Distributed under the [MIT](https://github.com/TimoBechtel/morfy/blob/main/LICENSE) License.

---

_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_


CHANGELOG.md
# 1.0.0 (2021-07-20)


### Bug Fixes

* **deps:** bump (security update) ([8b716db](https://github.com/TimoBechtel/morfy/commit/8b716dbf51ff95ebd510d836d23b68e62b4f8d8e))
* rename master to main to reflect branch renaming ([8e603ab](https://github.com/TimoBechtel/morfy/commit/8e603abfae703280182c42c12e2ab4172c4befb8))


### Features

* **typescript:** add typings for css ([ad96005](https://github.com/TimoBechtel/morfy/commit/ad960057df96c28306f1e9609e4db92608a183a5))
* add function to revert morhping ([23639c0](https://github.com/TimoBechtel/morfy/commit/23639c020c58235f016a32955bc7d3bb94786a0d))
* initial commit ([e2f6829](https://github.com/TimoBechtel/morfy/commit/e2f68290ef42bb90cf6c1a17227f23edf5141125))


src/index.js
const calcTransformation = (source, target) => {
  return {
    x: source.offsetLeft - target.offsetLeft,
    y: source.offsetTop - target.offsetTop,
    scaleY: target.offsetHeight / source.offsetHeight,
    scaleX: target.offsetWidth / source.offsetWidth,
  };
};

const getEffectedStyles = (element, properties) => {
  const computedStyle = window.getComputedStyle(element);
  return properties.reduce((styles, prop) => {
    styles[prop] = computedStyle[prop];
    return styles;
  }, {});
};

export const createMorphable = (
  source,
  target,
  {
    duration = 1,
    timingFunction = '',
    effectedCssProperties = ['background-color', 'border-radius', 'border'],
  } = {}
) => {
  const targetStyles = getEffectedStyles(target, effectedCssProperties);
  const sourceStyles = getEffectedStyles(source, effectedCssProperties);
  const transformation = calcTransformation(source, target);

  source.innerHTML = `<div>${source.innerHTML}</div>`;
  const contentWrapper = source.children[0];

  source.style.transformOrigin = 'top left';
  target.style.visibility = 'hidden';

  return {
    morph() {
      source.style.transition = `${['transform', ...effectedCssProperties].join(
        ` ${duration}s ${timingFunction},`
      )}, opacity ${duration}s ${duration * 1.5}s, visibility 0s ${
        duration * 1.5 + duration
      }s`;

      contentWrapper.style.transition = `opacity ${
        duration * 0.5
      }s ${timingFunction}`;

      target.style.transition = `opacity ${duration}s ${duration}s, visibility 0s ${duration}s`;

      source.style.transform = `translate(${-transformation.x}px, ${-transformation.y}px) scale(${
        transformation.scaleX
      }, ${transformation.scaleY})`;
      source.style.opacity = '0';
      source.style.visibility = 'hidden';

      effectedCssProperties.forEach(
        (prop) => (source.style[prop] = targetStyles[prop])
      );

      contentWrapper.style.opacity = '0';

      target.style.opacity = '1';
      target.style.visibility = 'visible';
    },
    revert() {
      source.style.transition = `${['transform', ...effectedCssProperties].join(
        ` ${duration}s ${duration * 0.2}s ${timingFunction},`
      )}, opacity 0s, visibility 0s`;

      contentWrapper.style.transition = `opacity ${duration}s ${
        duration * 0.5
      }s ${timingFunction}`;

      target.style.transition = `opacity ${duration * 0.2}s, visibility 0s ${
        duration * 0.2
      }s`;

      source.style.transform = `translate(0px, 0px) scale(1, 1)`;
      source.style.opacity = '1';
      source.style.visibility = 'visible';

      effectedCssProperties.forEach(
        (prop) => (source.style[prop] = sourceStyles[prop])
      );

      contentWrapper.style.opacity = '1';

      target.style.opacity = '0';
      target.style.visibility = 'hidden';
    },
  };
};

export const morph = (source, target, options) => {
  createMorphable(source, target, options).morph();
};


src/index.ts
import type { PropertiesHyphen } from 'csstype';

type StyleProperties = {
  [key in keyof PropertiesHyphen<string>]: string;
};

function calcTransformation(source: HTMLElement, target: HTMLElement) {
  return {
    x: source.offsetLeft - target.offsetLeft,
    y: source.offsetTop - target.offsetTop,
    scaleY: target.offsetHeight / source.offsetHeight,
    scaleX: target.offsetWidth / source.offsetWidth,
  };
}

function getAffectedStyles(
  element: HTMLElement,
  properties: (keyof StyleProperties)[]
): StyleProperties {
  const computedStyle = window.getComputedStyle(element);
  return properties.reduce((styles, prop) => {
    styles[prop] = computedStyle.getPropertyValue(prop);
    return styles;
  }, {} as StyleProperties);
}

export function createMorphable(
  source: HTMLElement,
  target: HTMLElement,
  {
    duration = 1,
    timingFunction = '',
    effectedCssProperties,
    affectedCssProperties = ['background-color', 'border-radius', 'border'],
  }: MorfyOptions = {}
): Morphable {
  // effectedCssProperties is deprecated, fallback to provide backwards compatibility
  if (effectedCssProperties) affectedCssProperties = effectedCssProperties;

  const targetStyles = getAffectedStyles(target, affectedCssProperties);
  const sourceStyles = getAffectedStyles(source, affectedCssProperties);
  const transformation = calcTransformation(source, target);

  source.innerHTML = `<div>${source.innerHTML}</div>`;
  const contentWrapper = source.children[0] as HTMLElement; // can be sure it is a HTMLElement as we created it before

  source.style.transformOrigin = 'top left';
  target.style.visibility = 'hidden';

  return {
    morph() {
      source.style.transition = `${['transform', ...affectedCssProperties].join(
        ` ${duration}s ${timingFunction},`
      )}, opacity ${duration}s ${duration * 1.5}s, visibility 0s ${
        duration * 1.5 + duration
      }s`;

      contentWrapper.style.transition = `opacity ${
        duration * 0.5
      }s ${timingFunction}`;

      target.style.transition = `opacity ${duration}s ${duration}s, visibility 0s ${duration}s`;

      source.style.transform = `translate(${-transformation.x}px, ${-transformation.y}px) scale(${
        transformation.scaleX
      }, ${transformation.scaleY})`;
      source.style.opacity = '0';
      source.style.visibility = 'hidden';

      affectedCssProperties.forEach(
        (prop) =>
          targetStyles[prop] !== undefined &&
          source.style.setProperty(prop, targetStyles[prop] as string)
      );

      contentWrapper.style.opacity = '0';

      target.style.opacity = '1';
      target.style.visibility = 'visible';
    },
    revert() {
      source.style.transition = `${['transform', ...affectedCssProperties].join(
        ` ${duration}s ${duration * 0.2}s ${timingFunction},`
      )}, opacity 0s, visibility 0s`;

      contentWrapper.style.transition = `opacity ${duration}s ${
        duration * 0.5
      }s ${timingFunction}`;

      target.style.transition = `opacity ${duration * 0.2}s, visibility 0s ${
        duration * 0.2
      }s`;

      source.style.transform = `translate(0px, 0px) scale(1, 1)`;
      source.style.opacity = '1';
      source.style.visibility = 'visible';

      affectedCssProperties.forEach(
        (prop) =>
          sourceStyles[prop] !== undefined &&
          source.style.setProperty(prop, sourceStyles[prop] as string)
      );

      contentWrapper.style.opacity = '1';

      target.style.opacity = '0';
      target.style.visibility = 'hidden';
    },
  };
}

export function morph(
  source: HTMLElement,
  target: HTMLElement,
  options: MorfyOptions
) {
  createMorphable(source, target, options).morph();
}

interface MorfyOptions {
  /**
   * duration in seconds
   */
  duration?: number;
  /**
   * css timing function
   * @example 'ease-in'
   */
  timingFunction?: PropertiesHyphen['transition-timing-function'];
  /**
   * @deprecated Will be removed in next major release. Use affectedCssProperties instead.
   */
  effectedCssProperties?: (keyof StyleProperties)[];

  /**
   * css properties to be transitioned from source to target
   */
  affectedCssProperties?: (keyof StyleProperties)[];
}

interface Morphable {
  /**
   * start morphing
   */
  morph: () => void;
  /**
   * morph to initial state
   */
  revert: () => void;
}
